File: webapp/codereview/internal/admin_service.py
# Copyright 2008 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import logging

from google.appengine.ext import db

from codereview import models
from codereview.models import gql

from admin_pb2 import AdminService
from sync_project_pb2 import *
from util import u, InternalAPI

class AdminServiceImp(AdminService, InternalAPI):
  def SyncProject(self, rpc_controller, req, done):
    rsp = SyncProjectResponse()

    proj = models.Project.get_project_for_name(req.project_name)
    if proj is None:
      proj = models.Project(name = req.project_name)
      proj.put()

    really_exists = set()
    for bs in req.branch:
      branch = models.Branch.get_or_insert_branch(proj, bs.branch_name)
      really_exists.add(branch.name)

    to_delete = []
    for b in list(gql(models.Branch, 'WHERE project = :1', proj)):
      if b.name not in really_exists:
        to_delete += gql(models.BuildAttempt, 'WHERE branch = :1', b)

    if to_delete:
      db.delete(to_delete)

    done(rsp)

File: appjar/src/main/java/com/google/gerrit/client/admin/AdminService.java
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.gerrit.client.admin;

import com.google.gerrit.client.reviewdb.AccountGroup;
import com.google.gerrit.client.reviewdb.AccountGroupMember;
import com.google.gerrit.client.rpc.SignInRequired;
import com.google.gwt.user.client.rpc.AsyncCallback;
import com.google.gwtjsonrpc.client.RemoteJsonService;
import com.google.gwtjsonrpc.client.VoidResult;

import java.util.Set;

public interface AdminService extends RemoteJsonService {
  @SignInRequired
  void groupDetail(AccountGroup.Id groupId,
      AsyncCallback<AccountGroupDetail> callback);

  @SignInRequired
  void changeGroupDescription(AccountGroup.Id groupId, String description,
      AsyncCallback<VoidResult> callback);

  @SignInRequired
  void addGroupMember(AccountGroup.Id groupId, String nameOrEmail,
      AsyncCallback<AccountGroupDetail> callback);

  @SignInRequired
  void deleteGroupMembers(Set<AccountGroupMember.Key> keys,
      AsyncCallback<VoidResult> callback);

  @SignInRequired
  void changeGroupOwner(AccountGroupMember.Key key, boolean owner,
      AsyncCallback<VoidResult> callback);
}
